File: divide2.c
#include <assert.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

int MIN = -2147483648;
int MAX = 2147483647;
int min(int a, int b) {
	if (a < b) {
	    return a;
	}
	return b;
}
int max(int a, int b) {
	if (a > b) {
	    return a;
	}
	return b;
}
int negate(int n) {
  if (n <= MIN) {
    printf("negate %d = MAX\n", n);
    return MAX;
  }
  if (n > MAX) {
    printf("negate %d = MIN\n", n);
    return MIN;
  }
    
  printf("negate %d = %d\n", n, -n);
  return -n;
}
bool optimize = true;
void div_and_check(int dividend, int divosor) {

  int expected = divosor == -1 ? negate(dividend) : dividend / divosor;
  expected = min(expected, MAX);
  expected = max(expected, MIN);
  assert(expected >= MIN);
  assert(expected <= MAX);
  int actual = (int)divide(dividend, divosor);
  if (actual != expected) {
    printf("%d / %d expected %d got %d\n", dividend, divosor, expected, actual);
    assert(actual == expected);
  } else {
    printf("PASS %d / %d expected %d got %d\n", dividend, divosor, expected, actual);
  }
}

int main(int argc, char **argv) {
  // div_and_check(10, 3);
  // div_and_check(1, 3);
  // div_and_check(9, 3);
  // div_and_check(9, 2);
  // div_and_check(9, 1);
  // div_and_check(9, -3);
  // div_and_check(9, 9);
  // div_and_check(0, -3);
  // div_and_check(MAX, MAX);
  // div_and_check(MIN, MIN);
  // div_and_check(MAX, 1);
  // div_and_check(9, 9);
  //div_and_check(MIN, -1);
  //div_and_check(MIN, 1);
  //div_and_check(2147483647, 1);
  //div_and_check(2147483647, 2);
  div_and_check(-2147483648, 2);
  //div_and_check(-2147483648, 1);
}

File: divide2.py
# vim:set makeprg=python
# import math
MIN = -2147483648
assert MIN == -(2**31)
MAX = 2147483647
assert MAX == 2**31 - 1
optimize = True

# def debug(unused ): # type: ignore
#    pass
MAX_INT = (1 << 31) - 1


def debug(s: str):
    print(s)


class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        debug(
            f"dividend {dividend} divisor {divisor} looking for {int(dividend/divisor)}"
        )
        assert divisor != 0
        if dividend == 0:
            return 0
        sign = 1 if ((dividend < 0) ^ (divisor < 0)) else 0
        dividend = abs(dividend)
        divisor = abs(divisor)
        res = 0
        while dividend >= divisor:
            debug(f"while dividend {dividend} >= divisor {divisor}")
            k = 0
            while dividend >= divisor << (k + 1):
                # debug(f"while dividend {dividend}  >= divisor {divisor} <<{k+1}  [{divisor<<(k+1)}]")
                k += 1
            dividend -= divisor << k
            debug(f"k {k} divisor<<k {divisor<<k} res+={1<<k}")
            # debug(f"subtract (divisor<<k) from divident = {dividend}")
            # debug(f"dividend {dividend} divisor {divisor} looking for {int(dividend/divisor)}")
            res += 1 << k
            debug(f"res {res}")
        return -res if sign else (res if res <= MAX_INT else MAX_INT)


solution = Solution()


def div(dend, dor):
    expected = int(dend / dor)
    expected = min(expected, MAX)
    expected = max(expected, MIN)
    assert expected >= MIN
    assert expected <= MAX
    actual = int(solution.divide(dend, dor))
    if actual != expected:
        print(f"{dend} / {dor} expected {expected} got {actual}")
        assert actual == expected
    else:
        print(f"PASS {dend} / {dor} = {expected} == {actual}")


if __name__ == "__main__":
    s = Solution()
    # div(10, 3)
    # div(1, 3)
    # div(9, 3)
    # div(9, 2)
    # div(9, 1)
    # div(9, -3)
    # div(9, 9)
    # div(0, -3)
    # div(MAX, MAX)
    # div(MIN, MIN)
    # div(MAX, 1)
    # div(9, 9)
    # div(MIN, -1)
    # div(MIN, 1)
    div(2147483647, 2)
